compiler/sem/semstmts.nim

```diff
@@ -1099,155 +1074,425 @@ proc semLetOrVar(c: PContext, n: PNode, symkind: TSymKind): PNode =
+        defPart.len > 3:
+      # xxx: helpful hints like this should likely be associated to an id/pos
+      #      instead of reporting
+      localReport(c.config, defPart.info, reportSem rsemEachIdentIsTuple)
+  else:
+    discard
 
   if hasError:
-    # xxx: hasError is here until we've converted the whole proc to use nkError
-    #      until then we check at the end to ensure we capture any nkErrors
-    #      embedded within the section
+    # wrap the result if there is an embedded error
     result = c.config.wrapError(result)
-
+    
   dec c.inStaticContext
 
 
+proc semConst(c: PContext, n: PNode): PNode =
+  ## semantically analyses const sections, analysis follows these steps:
+  ##
+  ## 0. results are accumulated into a statement list and holds the section
+  ## 1. for each entry in the section to analyse:
+  ##    - if a single ident in the identdef or vartuple it's analysed
+  ##    - multiple idents are decomposed and analysed one at a time
+  ## 2. analysis is first on the untyped version (macro pragma)
+  ## 3. typed analysis (interleaved with typed macro pragmas):
+  ##    1. expansion: rhs and lhs are analysed
+  ##    2. reduction: combined into an identdef, then final analysis
+  ##
+  ## If successful a singleton const section with a singleton constDef or
+  ## varTuple is produced, otherwise nkError. ConstDef's name part is
+  ## normalized to always contain have a pragmaExprs, even with an empty
+  ## pragmas section.
+
+  # xxx - likely mergeable with proc `semVarOrLetSection`
+
+  addInNimDebugUtils(c.config, "semConst", n, result)
+
+  # initial basic validation
+
+  assert n != nil, "nil node instead of a const section"
+  internalAssert(c.config, n.kind in {nkConstSection},
+    "only const sections allowed, got: " & $n.kind)
+  
+  # setup for untyped pragma processing
+
+  result = newNodeI(nkStmtList, n.info) # accumulate the result here
+
+  for i, a in n.pairs:
+    const allowedNodeKinds = {nkConstDef, nkVarTuple}
+      ## valid definitions allowed in let or var sections
+
+    if c.config.cmd == cmdIdeTools:
+      suggestStmt(c, a)
+
+    case a.kind
+    of nkCommentStmt: continue # skip comments
+    of nkConstDef:
+      checkMinSonsLen(a, 3, c.config)
+      
+      for j in 0..<a.len - 2:
+        # assemble a var section per ident defined
+        let
+          info = a[j].info
+          singletonNode = copyNode(n)
+          singletonDef = newNodeI(nkConstDef, info, 3)
+
+        singletonNode.info = info # set the info to the ident start
+
+        if importantComments(c.config):
+          # keep documentation information:
+          singletonDef.comment = a.comment
+
+        singletonDef[0] = a[j]
+        singletonDef[1] = copyTree(a[^2])
+        singletonDef[2] = copyTree(a[^1])
+
+        singletonNode.add singletonDef
+
+        # process pragmas and based on that direct further analysis
+        let pragmad = semConstLetOrVarAnnotation(c, singletonNode)
```

Comment:
The name `pragmad` is a bit confusing. I had to read it multiple times in order to realize that it's likely supposed to mean "def with pragmas applied".

My comment isn't meant as a request for it to change - just as feedback.

Reply:
Thanks, I added a comment to help, maybe a better name will become evident soon.